kotlin-jvm/src/main/kotlin/playground/di/Koin.kt

```diff
@@ -0,0 +1,102 @@
+@file:Suppress("PackageDirectoryMismatch")
+
+package playground.di.koin
+
+import org.koin.core.KoinComponent
+import org.koin.core.context.startKoin
+import org.koin.core.inject
+import org.koin.core.qualifier.named
+import org.koin.dsl.bind
+import org.koin.dsl.module
+import playground.shouldBe
+
+/**
+ * Koin: A pragmatic lightweight dependency injection framework for Kotlin developers.
+ *
+ * - [Website](https://insert-koin.io/)
+ * - [Github](https://github.com/InsertKoinIO/koin)
+ * - [CHANGELOG](https://github.com/InsertKoinIO/koin/blob/master/CHANGELOG.md)
+ */
+fun main() {
+    val deps = module {
+        single(qualifier = named("production")) {
+            InMemoryUserStore()
+        } bind UserRepository::class
+        single(qualifier = named("test")) {
+            MockedUserStore()
+        } bind UserRepository::class
+    }
+
+    startKoin {
+        modules(deps)
+    }
+
+    val user = User(1)
+
+    val testApp = Application("test")
+    val realApp = Application("production")
+
+    testApp.addUser(user)
+    // In the test app, we want always the same user.
+    testApp.getUser(1) shouldBe User(123)
+
+    realApp.addUser(user)
+    // IN the real app, we want the users we have previously stored.
+    realApp.getUser(1) shouldBe user
+}
+
+
+/**
+ * Abstraction of an application that can get and add users given a  UserRepository.
+ */
+internal class Application(environment: String) : KoinComponent {
+    private val userRepository by inject<UserRepository>(qualifier = named(environment))
```

Comment:
I don't know about Koin, but in general I prefer to use constructor injection when that's possible. 
Is that an option here?

Reply:
I tried figuring out a way of using the constructor injector, but with Koin it's not as easy to achieve it as with KodeinDI :/

It requires at least 2 extra classes to achieve this. What do you think?

Reply:
Ok, leave it as is for now!